src/net/ossl.cc

```diff
@@ -453,5 +462,592 @@ void tls::server_credentials::set_client_auth(client_auth ca) {
+        });
+    }
+
+    future<> do_shutdown() {
+        netlogger.info("Inside session::do_shutdown");
+        /// TODO(rob) - Is yield the right option here
+        if(_error || !connected()) {
+            return make_ready_future();
+        }
+        auto res = SSL_shutdown(_ssl.get());
+        if (res == 1){
+            // Shutdown has completed successfully
+            netlogger.info("Session shutdown gracefully");
+            return make_ready_future<>();
+        } else if (res == 0) {
+            // Shutdown process is ongoing and has not yet completed, peer has not yet replied
+            return yield().then([this](){
+                return do_shutdown();
+            });
+        }
+        // Shutdown was not successful
+        auto f = make_ready_future<>();
+        auto err = SSL_get_error(_ssl.get(), res);
+        if (err == SSL_ERROR_WANT_READ || err == SSL_ERROR_WANT_WRITE){
+            if (err == SSL_ERROR_WANT_READ) {
+                f = wait_for_input();
+            }
+            return f.then([]{ return yield(); }).then([this]{
+                if (!eof()) {
+                    return do_shutdown();
+                }
+                return make_ready_future<>();
+            });
+        }
+
+        // Fatal error
+        return make_exception_future<>(ossl_error(res, "fatal error during ssl shutdown"));
+    }
+
+    template<typename Fn>
+    static std::optional<sstring> get_ossl_string(Fn f, const x509_ptr& cert){
+        X509_NAME *name = f(cert.get());
+        if (name == nullptr) {
+            return std::nullopt;
+        }
+        char *name_str = X509_NAME_oneline(name, NULL, 0);
+        if (name_str == nullptr) {
+            return std::nullopt;
+        }
+        sstring ossl_str(name_str);
+        OPENSSL_free(name_str);
+        return ossl_str;
+    }
+
+    void verify() {
+        auto res = SSL_get_verify_result(_ssl.get());
+        if (res != X509_V_OK) {
+            sstring stat_str(X509_verify_cert_error_string(res));
+            auto dn = extract_dn_information();
+            if (dn) {
+                std::stringstream ss;
+                ss << stat_str;
+                if (stat_str.back() != ' ') {
+                    ss << ' ';
+                }
+                ss << "(Issuer=[" << dn->issuer << "], Subject=[" << dn->subject << "])";
+                stat_str = ss.str();
+            }
+            throw verification_error(stat_str);
+        }
+        // A success return code does not signify if a cert was presented or not, that must be explicitly
+        // queried via SSL_get_peer_cert
+        auto peer_cert = SSL_get_peer_certificate(_ssl.get());
+        if(peer_cert == nullptr && _type == type::SERVER){
+            if(_creds->get_client_auth() == client_auth::REQUIRE) {
+                // No peer certificate was returned even though it was requested and required
+                throw ossl_error(res, "Error within SSL verify");
```

Comment:
nit: I think this error message should be a little bit more verbose stating why there was an error